Mathlib/Logic/Equiv/Basic.lean

```diff
@@ -0,0 +1,1811 @@
+  equivPempty _
+#align equiv.prod_pempty Equiv.prodPempty
+
+/-- `pempty` type is a left absorbing element for type product up to an equivalence. -/
+def pemptyProd (α : Type _) : PEmpty × α ≃ PEmpty :=
+  equivPempty _
+#align equiv.pempty_prod Equiv.pemptyProd
+
+end
+
+section
+
+open Sum
+
+/-- `psum` is equivalent to `sum`. -/
+def psumEquivSum (α β : Type _) : PSum α β ≃ Sum α β where
+  toFun s := PSum.casesOn s inl inr
+  invFun := Sum.elim PSum.inl PSum.inr
+  left_inv s := by cases s <;> rfl
+  right_inv s := by cases s <;> rfl
+#align equiv.psum_equiv_sum Equiv.psumEquivSum
+
+/-- If `α ≃ α'` and `β ≃ β'`, then `α ⊕ β ≃ α' ⊕ β'`. This is `sum.map` as an equivalence. -/
+@[simps apply]
+def sumCongr {α₁ β₁ α₂ β₂ : Type _} (ea : α₁ ≃ α₂) (eb : β₁ ≃ β₂) : Sum α₁ β₁ ≃ Sum α₂ β₂ :=
+  ⟨Sum.map ea eb, Sum.map ea.symm eb.symm, fun x => by simp, fun x => by simp⟩
+#align equiv.sum_congr Equiv.sumCongr
+
+/-- If `α ≃ α'` and `β ≃ β'`, then `psum α β ≃ psum α' β'`. -/
+def psumCongr {δ : Sort z} (e₁ : α ≃ β) (e₂ : γ ≃ δ) : PSum α γ ≃ PSum β δ where
+  toFun x := PSum.casesOn x (PSum.inl ∘ e₁) (PSum.inr ∘ e₂)
+  invFun x := PSum.casesOn x (PSum.inl ∘ e₁.symm) (PSum.inr ∘ e₂.symm)
+  left_inv := by rintro (x | x) <;> simp
+  right_inv := by rintro (x | x) <;> simp
+#align equiv.psum_congr Equiv.psumCongr
+
+/-- Combine two `equiv`s using `psum` in the domain and `sum` in the codomain. -/
+def psumSum {α₁ β₁ : Sort _} {α₂ β₂ : Type _} (ea : α₁ ≃ α₂) (eb : β₁ ≃ β₂) : PSum α₁ β₁ ≃ Sum α₂ β₂ :=
+  (ea.psumCongr eb).trans (psumEquivSum _ _)
+#align equiv.psum_sum Equiv.psumSum
+
+/-- Combine two `equiv`s using `sum` in the domain and `psum` in the codomain. -/
+def sumPsum {α₁ β₁ : Type _} {α₂ β₂ : Sort _} (ea : α₁ ≃ α₂) (eb : β₁ ≃ β₂) : Sum α₁ β₁ ≃ PSum α₂ β₂ :=
+  (ea.symm.psumSum eb.symm).symm
+#align equiv.sum_psum Equiv.sumPsum
+
+@[simp]
+theorem sum_congr_trans {α₁ α₂ β₁ β₂ γ₁ γ₂ : Sort _} (e : α₁ ≃ β₁) (f : α₂ ≃ β₂) (g : β₁ ≃ γ₁) (h : β₂ ≃ γ₂) :
+    (Equiv.sumCongr e f).trans (Equiv.sumCongr g h) = Equiv.sumCongr (e.trans g) (f.trans h) := by
+  ext i
+  cases i <;> rfl
+#align equiv.sum_congr_trans Equiv.sum_congr_trans
+
+@[simp]
+theorem sum_congr_symm {α β γ δ : Sort _} (e : α ≃ β) (f : γ ≃ δ) :
+    (Equiv.sumCongr e f).symm = Equiv.sumCongr e.symm f.symm :=
+  rfl
+#align equiv.sum_congr_symm Equiv.sum_congr_symm
+
+@[simp]
+theorem sum_congr_refl {α β : Sort _} : Equiv.sumCongr (Equiv.refl α) (Equiv.refl β) = Equiv.refl (Sum α β) := by
+  ext i
+  cases i <;> rfl
+#align equiv.sum_congr_refl Equiv.sum_congr_refl
+
+namespace Perm
+
+/-- Combine a permutation of `α` and of `β` into a permutation of `α ⊕ β`. -/
+@[reducible]
+def sumCongr {α β : Type _} (ea : Equiv.Perm α) (eb : Equiv.Perm β) : Equiv.Perm (Sum α β) :=
+  Equiv.sumCongr ea eb
+#align equiv.perm.sum_congr Equiv.Perm.sumCongr
+
+@[simp]
+theorem sum_congr_apply {α β : Type _} (ea : Equiv.Perm α) (eb : Equiv.Perm β) (x : Sum α β) :
```

Comment:
Should be named `sumCongr_apply`. Make sure the update the corresponding #align at the same time.

This affects many definitions elsewhere in the file.